File: examples/csharp/Patterns/Decorator.cs
namespace DesignPatterns.Decorator
{
    public interface IText
    {
        string GetContent();
    }

    public class PlainText : IText
    {
        private readonly string _text;
        public PlainText(string text) => _text = text;
        public string GetContent() => _text;
    }

    public abstract class TextDecorator : IText
    {
        protected readonly IText Wrapped;
        protected TextDecorator(IText wrapped) => Wrapped = wrapped;
        public abstract string GetContent();
    }

    public class BoldDecorator : TextDecorator
    {
        public BoldDecorator(IText wrapped) : base(wrapped) { }
        public override string GetContent() => $"<b>{Wrapped.GetContent()}</b>";
    }

    public class ItalicDecorator : TextDecorator
    {
        public ItalicDecorator(IText wrapped) : base(wrapped) { }
        public override string GetContent() => $"<i>{Wrapped.GetContent()}</i>";
    }

    public class UnderlineDecorator : TextDecorator
    {
        public UnderlineDecorator(IText wrapped) : base(wrapped) { }
        public override string GetContent() => $"<u>{Wrapped.GetContent()}</u>";
    }

    public static class Demo
    {
        public static void Run()
        {
            IText text = new PlainText("Hello, World!");
            Console.WriteLine($"Plain:       {text.GetContent()}");

            text = new BoldDecorator(new PlainText("Hello, World!"));
            Console.WriteLine($"Bold:        {text.GetContent()}");

            text = new ItalicDecorator(new BoldDecorator(new PlainText("Hello, World!")));
            Console.WriteLine($"Bold+Italic: {text.GetContent()}");

            text = new UnderlineDecorator(
                new ItalicDecorator(new BoldDecorator(new PlainText("Hello, World!"))));
            Console.WriteLine($"All three:   {text.GetContent()}");
        }
    }
}

File: examples/cpp/decorator.cpp
#include <iostream>
#include <memory>
#include <string>

class Text {
public:
    virtual ~Text() = default;
    virtual std::string render() const = 0;
};

class PlainText : public Text {
    std::string content_;
public:
    PlainText(const std::string& content) : content_(content) {}
    std::string render() const override { return content_; }
};

class TextDecorator : public Text {
protected:
    std::unique_ptr<Text> wrapped_;
public:
    TextDecorator(std::unique_ptr<Text> text) : wrapped_(std::move(text)) {}
};

class BoldDecorator : public TextDecorator {
public:
    using TextDecorator::TextDecorator;
    std::string render() const override {
        return "<b>" + wrapped_->render() + "</b>";
    }
};

class ItalicDecorator : public TextDecorator {
public:
    using TextDecorator::TextDecorator;
    std::string render() const override {
        return "<i>" + wrapped_->render() + "</i>";
    }
};

int main() {
    std::cout << "=== Decorator: Text Formatting ===" << std::endl;

    auto plain = std::make_unique<PlainText>("Hello, World!");
    std::cout << "Plain: " << plain->render() << std::endl;

    auto bold = std::make_unique<BoldDecorator>(std::make_unique<PlainText>("Hello, World!"));
    std::cout << "Bold: " << bold->render() << std::endl;

    auto boldItalic = std::make_unique<ItalicDecorator>(
        std::make_unique<BoldDecorator>(std::make_unique<PlainText>("Hello, World!")));
    std::cout << "Bold+Italic: " << boldItalic->render() << std::endl;

    return 0;
}
